File: problem/Min_Total_Distance_Traveled/minimumTotalDistance.py
class Solution:
    def minimumTotalDistance(self, robot, factory):
        robot.sort()
        factory.sort()
        
        n = len(robot)
        m = len(factory)
        dp = [[-1] * m for _ in range(n)]

        def solve(i, j):
            if i == n:
                return 0
            if j == m:
                return int(1e15)
            if dp[i][j] != -1:
                return dp[i][j]

            res = solve(i, j + 1)
            cost = 0
            pos = factory[j][0]
            limit = factory[j][1]

            for k in range(limit):
                if i + k >= n:
                    break
                cost += abs(robot[i + k] - pos)
                res = min(res, cost + solve(i + k + 1, j + 1))

            dp[i][j] = res
            return res

        return solve(0, 0)

File: problem/Min_Total_Distance_Traveled/minimumTotalDistance.java
import java.util.*;

class Solution {
    private long[][] dp;

    public long minimumTotalDistance(List<Integer> robot, int[][] factory) {
        Collections.sort(robot);
        Arrays.sort(factory, (a, b) -> Integer.compare(a[0], b[0]));

        int n = robot.size();
        int m = factory.length;
        dp = new long[n][m];
        for (long[] row : dp) {
            Arrays.fill(row, -1);
        }

        return solve(0, 0, robot, factory);
    }

    private long solve(int i, int j, List<Integer> robot, int[][] factory) {
        int n = robot.size();
        int m = factory.length;

        if (i == n) return 0;
        if (j == m) return (long) 1e15;
        if (dp[i][j] != -1) return dp[i][j];

        long res = solve(i, j + 1, robot, factory);

        long cost = 0;
        int pos = factory[j][0];
        int limit = factory[j][1];

        for (int k = 0; k < limit && i + k < n; k++) {
            cost += Math.abs((long) robot.get(i + k) - pos);
            long subproblem = solve(i + k + 1, j + 1, robot, factory);
            if (subproblem != (long) 1e15) {
                res = Math.min(res, cost + subproblem);
            }
        }

        return dp[i][j] = res;
    }
}